### PR TITLE
fix(GODT-1896): Do not remove subscribed status for deleted mailboxes

### PR DESCRIPTION
**Pull request by LBeernaertProton**
Ensure that we do not delete subscriptions entries for deleted mailboxes without an explicit unsubscribe event (RFC 3501 Section 6.3.9).

If a new mailbox is created with the same name as previously deleted and subscribed mailbox, the subscription data is overwritten to point to the newly created mailbox.

### CODE REVIEW

File: internal/state/match.go [hunk 1]
```diff
@@ -3,11 +3,11 @@ package state
 import (
 	"context"
 	"fmt"
+	"github.com/ProtonMail/gluon/internal/db"
```
[the remark this answers is not in the document]
Comment (LBeernaertProton) [reply internal/state/match.go]: Done

File: internal/db/subscription.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import (
+	"context"
+	"github.com/ProtonMail/gluon/imap"
+	"github.com/ProtonMail/gluon/internal/db/ent"
+	"github.com/ProtonMail/gluon/internal/db/ent/subscription"
+	"github.com/ProtonMail/gluon/internal/ids"
+)
+
+func AddSubscription(ctx context.Context, tx *ent.Tx, mboxName string, mboxID ids.MailboxIDPair) error {
+	count, err := tx.Subscription.Update().Where(subscription.NameEqualFold(mboxName)).SetMailboxID(mboxID.InternalID).SetRemoteID(mboxID.RemoteID).Save(ctx)
```
Comment (jameshoulahan) [inline internal/db/subscription.go:12]: I'm happy with how you did this, no need to change it, but just fyi we might be able to use upsert here (https://entgo.io/docs/feature-flags#upsert)

Comment (LBeernaertProton) [reply internal/db/subscription.go:12]: Noted!

**Comment by LBeernaertProton** [comment]
> 

Good catch.  Fixed and unit test added.

File: internal/state/match.go [hunk 3]
```diff
@@ -18,16 +18,23 @@ type Match struct {
 	Atts      imap.FlagSet
 }
 
+type MatchMailbox struct {
```
Comment (jameshoulahan) [inline internal/state/match.go:21]: Could you make this private (`matchMailbox`) so as to not confuse with the other public "pseudo-mailbox" (`AppendOnlyMailbox`)?

Comment (LBeernaertProton) [reply internal/state/match.go]: Done

**Review by jameshoulahan — APPROVED** [review]

**Pull request merged**